### PR TITLE
WIP: vscript support

### PR DESCRIPTION
**Pull request by ReDucTor**
Starting this PR to start the discussion/review while some remaining parts are left over

Remaining parts
- [x] FGD changes (https://github.com/mapbase-source/source-fgds/pull/1)
- [x] Documentation/Wiki related changes
- [ ] Identify and document potential incompatibilities with ASW/CS:GO/Portal2 vscripts
- [x] Save and restore support
- [ ] Easier primitive types

### CODE REVIEW

File: sp/src/game/server/player.cpp [hunk 1]
```diff
@@ -5300,6 +5314,14 @@ void CBasePlayer::OnRestore( void )
 	m_nVehicleViewSavedFrame = 0;
 
 	m_nBodyPitchPoseParam = LookupPoseParameter( "body_pitch" );
+
+	// HACK: (03/25/09) Then the player goes across a transition it doesn't spawn and register
+	// it's instance. We're hacking around this for now, but this will go away when we get around to 
+	// having entities cross transitions and keep their script state.
+	if ( !g_pGameRules->IsMultiplayer() && g_pScriptVM && (gpGlobals->eLoadType == MapLoad_Transition) )
```
Comment (Blixibon) [inline sp/src/game/server/player.cpp:5321]: Will proper save/restore support for VScripts fix this issue?

File: sp/src/vpc_scripts/source_base.vpc [hunk 2]
```diff
@@ -21,6 +21,9 @@ $Conditional MAPBASE	"1"
 
 // Toggles Mapbase's RPC implementation
 $Conditional MAPBASE_RPC	"1"
+
+// Toggles VScript implementation (note: interfaces still exist, just the provided implementation is not present)
+$Conditional MAPBASE_VSCRIPT	"1"
```
Comment (Blixibon) [inline sp/src/vpc_scripts/source_base.vpc:26]: Thank you for adding a unique VPC conditional and C++ preprocessor for this. I will probably use it for any Mapbase-specific expansions or changes to the VScript implementation unless you intended it to be used for more specific purposes only. (e.g. how it's used in `vscript_shared.cpp`)

File: sp/src/game/server/baseentity.cpp [hunk 3]
```diff
@@ -1322,17 +1340,16 @@ void CBaseEntity::FireNamedOutput( const char *pszOutput, variant_t variant, CBa
 			typedescription_t *dataDesc = &dmap->dataDesc[i];
 			if ( ( dataDesc->fieldType == FIELD_CUSTOM ) && ( dataDesc->flags & FTYPEDESC_OUTPUT ) )
 			{
-				CBaseEntityOutput *pOutput = ( CBaseEntityOutput * )( ( int )this + ( int )dataDesc->fieldOffset[0] );
+				CBaseEntityOutput *pOutput = ( CBaseEntityOutput * )( ( int )this + ( int )dataDesc->fieldOffset );
```
[the remark this answers is not in the document]
Comment (ReDucTor) [reply sp/src/game/server/baseentity.cpp:1343]: Ya, looks like this was a bad copy and paste from Alien Swarm, it doesn't have packed `fieldOffset`'s just single ones

File: sp/src/game/server/player.cpp (hunk 1 above)
Comment (ReDucTor) [reply sp/src/game/server/player.cpp:5321]: This is a hack from Alien Swarm, not something that save/restore will impact.

File: sp/src/game/server/baseentity.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ReDucTor) [reply sp/src/game/server/baseentity.cpp]: Resolved with d52196a0

File: sp/src/game/server/logicentities.cpp [hunk 4]
```diff
@@ -27,6 +27,126 @@
...
+	DEFINE_KEYFIELD( m_iszGroupMembers[6], FIELD_STRING, "Group06"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[7], FIELD_STRING, "Group07"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[8], FIELD_STRING, "Group08"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[9], FIELD_STRING, "Group09"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[10], FIELD_STRING, "Group10"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[11], FIELD_STRING, "Group11"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[12], FIELD_STRING, "Group12"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[13], FIELD_STRING, "Group13"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[14], FIELD_STRING, "Group14"),
+	DEFINE_KEYFIELD( m_iszGroupMembers[15], FIELD_STRING, "Group16"),
```
Comment (ReDucTor) [inline sp/src/game/server/logicentities.cpp:145]: The field name Group16 seems out of sync should be Group15, its copy/paste from ASW, but possibly should be fixed, these don't appear in the ASW FGD so can't see if they got it right in there, however looking other vscript FGD's it uses Group15 not Group16

Comment (ReDucTor) [reply sp/src/game/server/logicentities.cpp]: Resolved in 3ad34f3

**Comment by TeamSpen210** [comment]
Just so you know, in Valve's implementation regexes aren't save-loadable, and the `point_template` logic you were editing isn't fully functional.

**Pull request merged**